### PR TITLE
#25 providing only unassigned ports

### LINKED ISSUE #25: Assign only free ports
When the Node app is restarted, it starts to assign ports already taken by other build instances. It should firstly check whether particular port has not been taken yet.

### PR DESCRIPTION
**Pull request by Kryniol**
fixes #25

### CODE REVIEW

File: server/modules/instantiation/jobs.js [hunk 1]
```diff
@@ -303,16 +305,19 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
             var { buildInstance } = job;
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: not visible in pr but:
```
supports(job => job instanceif ProvidePortJob)
```

File: server/modules/instantiation/jobs.js [hunk 2]
```diff
@@ -303,16 +305,19 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
 
             return new Promise((resolve) => {
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]:
```
return new Promise(resolve => {
```
(redundant () )

File: server/modules/instantiation/jobs.js [hunk 3]
```diff
@@ -303,16 +305,19 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
 
-                buildInstanceRepository
-                    .updateExternalPorts(buildInstance)
-                    .then(resolve);
+                    buildInstance.log(`Assigned port ${port} for ${job.portName} external port.`);
+
+                    buildInstanceRepository
+                        .updateExternalPorts(buildInstance)
+                        .then(resolve);
+                });
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]:
```
(new PortProvider())
.providePort(_.map(job.portRanges, portRange => new PortRange(portRange[0], portRange[1])))
.then(port => {
  ...
});
```
formatting (new line with dot, redundant assigment etc.)

File: server/modules/instantiation/port-provider.js [hunk 4]
```diff
@@ -52,28 +57,31 @@ module.exports = function () {
...
-                        this.addExternalPortRange(port, port);
-
-                        return port;
-                    }
-                }
-            }
-
-            throw new Error('Failed to allocate port.');
+            return Promise
+                .any(expectedPortRanges.map(portRange => this.findPort(portRange)))
```
Comment (wit3k) [inline server/modules/instantiation/port-provider.js]:
```
.any(expectedPortRanges.map(this.findPort))
```
should work as well

File: server/modules/instantiation/port-provider.js [hunk 5]
```diff
@@ -52,28 +57,31 @@ module.exports = function () {
...
+        findPort(portRange) {
+            portfinder.basePort = portRange.minPort;
+
+            return new Promise((resolve, reject) => {
+                portfinder.getPortPromise()
+                    .then(port => {
+                        if (this.availablePortRanges.contains(port) && portRange.contains(port)) {
+                            resolve(port);
+
+                            return;
```
Comment (wit3k) [inline server/modules/instantiation/port-provider.js]: not sure, but this return is probably redundant

File: server/modules/instantiation/port-provider.js (hunk 4 above)
Comment (Kryniol) [inline server/modules/instantiation/port-provider.js]: initially I thought so as well
this actually was the biggest struggle here and after all we found out this changes the context and `this` becomes something else than it used to be

File: server/modules/instantiation/port-provider.js (hunk 5 above)
Comment (wit3k) [inline server/modules/instantiation/port-provider.js]: and if so this probably could become:
```
return new Promise((resolve, reject) =>
  portfinder
  .getPortPromise()
  .then(port =>
    this.availablePortRanges.contains(port) && portRange.contains(port)
      ? resolve(port)
      : reject(new Error('Couldn\'t find any available unallocated port in given range')))
);
```

Comment (Kryniol) [inline server/modules/instantiation/port-provider.js]: this was another big struggle here, apparently `resolve()` doesn't terminate the execution so actually it is needed

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: the formatting could be simplified in the whole file, let's stick to what I've changed here, jobs.js will be refactored to smaller classes anyway so then we can improve the formatting a bit

File: server/modules/instantiation/port-provider.js (hunk 5 above)
Comment (wit3k) [inline server/modules/instantiation/port-provider.js]: ok

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: ohh comon it's just one line

File: server/modules/instantiation/port-provider.js (hunk 4 above)
Comment (wit3k) [inline server/modules/instantiation/port-provider.js]: ok

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: your version doesn't work :(

Comment (wit3k) [inline server/modules/instantiation/jobs.js]: :/

**Comment by wit3k** [comment]
:+1: but I see that there is no travis setup for this project :/
You probably already thought about it, so I'm just reminiding

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (malef) [inline server/modules/instantiation/jobs.js]: Probably should have been `instanceof` :smile:

Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: nah, I didn't literally copy it, tried the one without any typos and it didn't work :)

**Pull request merged**